lib/cli/watch-run.js

```diff
@@ -17,14 +17,42 @@ const collectFiles = require('./collect-files');
  * @param {Mocha} mocha - Mocha instance
  * @param {Object} opts - Options
  * @param {string} opts.ui - User interface
+ * @param {string[]} [opts.watchFiles] - List of paths and patterns to
+ *   watch. If not provided all files with an extension included in
+ *   `fileColletionParams.extension` are watched. See first argument of
+ *   `chokidar.watch`.
+ * @param {string[]} opts.watchIgnore - List of paths and patterns to
+ *   exclude from watching. See `ignored` option of `chokidar`.
  * @param {Object} fileCollectParams - Parameters that control test
  *   file collection. See `lib/cli/collect-files.js`.
- * @param {string[]} fileCollectParams.extension - List of extensions to watch
+ * @param {string[]} fileCollectParams.extension - List of extensions
+ *   to watch if `opts.watchFiles` is not given.
  * @private
  */
-module.exports = (mocha, {ui}, fileCollectParams) => {
-  let runner;
-  const files = collectFiles(fileCollectParams);
+module.exports = (mocha, {ui, watchFiles, watchIgnore}, fileCollectParams) => {
```

Comment:
My IDE warns: 'ui' is declared but its value is never read.
All three properties of `obj` are already enclosed in the `mocha`instance. Do we need to pass this `obj` at all?

Reply:
Indeed, `ui` is unused. I removed it in https://github.com/mochajs/mocha/pull/3980/commits/10f4fda066ca96a242a3a69d52207e70664f1a01.

For the other options I’d favor passing them explicitly. They are only available through `mocha.options` because we dump all options there but it’s not clear whether this is an internal property or not. I tried isolate all the access to internal mocha properties in `resetMocha` (which should probably be turned into a method at some point).